File: benchmarks/agent-scorecard/typescript/tool-memory-workflow.ts
type AuthoritySlot = { effect: string; consumed: boolean };

function consume(slot: AuthoritySlot, effect: string): void {
  if (slot.consumed || slot.effect !== effect) throw new Error("authority rejected");
  slot.consumed = true;
}

const apiOnce: AuthoritySlot = { effect: "api_call", consumed: false };
const memoryOnce: AuthoritySlot = { effect: "memory_write", consumed: false };
const result: bigint = 6n * 7n;
consume(apiOnce, "api_call");
consume(memoryOnce, "memory_write");
process.stdout.write(`${JSON.stringify({
  effect_requests: [
    {
      arguments: [{ type: "string", value: "model" }, { type: "i64", value: result.toString() }],
      authority_slot: "api_once",
      effect: "api_call",
      information: { class: "public" },
    },
    {
      arguments: [{ type: "string", value: "session" }, { type: "i64", value: result.toString() }],
      authority_slot: "memory_once",
      effect: "memory_write",
      information: { class: "public" },
    },
  ],
  result: { type: "i64", value: result.toString() },
})}\n`);

File: benchmarks/agent-scorecard/rust/tool-memory-workflow.rs
struct AuthoritySlot {
    effect: &'static str,
    consumed: bool,
}

impl AuthoritySlot {
    fn consume(&mut self, effect: &str) -> Result<(), &'static str> {
        if self.consumed || self.effect != effect {
            return Err("authority rejected");
        }
        self.consumed = true;
        Ok(())
    }
}

fn main() -> Result<(), &'static str> {
    let mut api_once = AuthoritySlot { effect: "api_call", consumed: false };
    let mut memory_once = AuthoritySlot { effect: "memory_write", consumed: false };
    let result = 6_i64 * 7_i64;
    api_once.consume("api_call")?;
    memory_once.consume("memory_write")?;
    println!("{{\"effect_requests\":[{{\"arguments\":[{{\"type\":\"string\",\"value\":\"model\"}},{{\"type\":\"i64\",\"value\":\"{result}\"}}],\"authority_slot\":\"api_once\",\"effect\":\"api_call\",\"information\":{{\"class\":\"public\"}}}},{{\"arguments\":[{{\"type\":\"string\",\"value\":\"session\"}},{{\"type\":\"i64\",\"value\":\"{result}\"}}],\"authority_slot\":\"memory_once\",\"effect\":\"memory_write\",\"information\":{{\"class\":\"public\"}}}}],\"result\":{{\"type\":\"i64\",\"value\":\"{result}\"}}}}");
    Ok(())
}
